Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -491,63 +497,67 @@ private func createResolvedPackages(
-        let curPkg = builder.package.manifest.displayName
-        builder.package.targets.forEach { target in
-            if let parentPkgs = pkgChainMap[curPkg] {
-                // If a parent (downstream) package contains a module alias for this target
-                // from curPkg, add that alias and curPkg to the map
-                let elements = targetNameToAliasAndPkgMap.map { (targetName, aliasAndPkgList) -> [(String, String)] in
-                    if target.name == targetName {
-                        let tuples = aliasAndPkgList.filter { parentPkgs.contains($0.1) && !$0.0.isEmpty }
-                        return tuples.map { ($0.0, curPkg)}
-                    }
-                    return []
-                }.flatMap {$0}
-                targetNameToAliasAndPkgMap[target.name, default: []].append(contentsOf: elements)
-            }
+    // There could be multiple root packages but the common cases involve
+    // just one root package; handling multiple roots is tracked rdar://88518683
+    if let rootPkg = rootPkgID {
+        var pkgStack = [PackageIdentity]()
+        walkPkgTreeAndGetModuleAliases(for: rootPkg, with: packagesByIdentity, using: &pkgStack, output: &result)
+    }
+    return result
+}
 
+// Walk a package dependency tree and set the module aliases for targets in each package.
+// If multiple aliases are specified in upstream packages, aliases specified most downstream
+// will be used.
+private func walkPkgTreeAndGetModuleAliases(for pkgID: PackageIdentity,
+                                             with packagesByIdentity: [PackageIdentity: ResolvedPackageBuilder],
+                                             using pkgStack: inout [PackageIdentity],
+                                             output result: inout [PackageIdentity: [String: String]]) {
+    // Get the builder first
+    if let builder = packagesByIdentity[pkgID] {
+        // Add the pkgID to a stack used to keep track of
+        // multiple aliases specified in the package chain
+        pkgStack.append(pkgID)
+        builder.package.targets.forEach { target in
             target.dependencies.forEach { dep in
+                // Check if a dependency for this target has module aliases specified
                 if case let .product(prodRef, _) = dep {
                     if let prodPkg = prodRef.package {
                         if let prodModuleAliases = prodRef.moduleAliases {
                             for (depName, depAlias) in prodModuleAliases {
-                                // Add a module alias specified for this target's dependency
-                                targetNameToAliasAndPkgMap[depName, default: []].append((depAlias, prodPkg))
+                                let prodPkgID = PackageIdentity.plain(prodPkg)
+                                // Add the specified alias and the dependency package to a map
+                                result[prodPkgID] = [depName: depAlias]
```

Comment:
I might be missing something, but doesn't this need to append to the existing dictionary if `result[prodPkgID]` is non-nil?

Reply:
There should be only one module alias for each module from the package prodPkgID; if there are multiple aliases for a module from the same package, we should throw an error. I'll add a note for this error handling.